### PR TITLE
Control of random generator

### PR DESCRIPTION
**Pull request by jpaillard**
- For CFI and PFI, move `rng = check_random_state(self.random_state)` to the predict/importance stage to facilitate generating deterministic results 
 - For ConditionalSampling move it to the `.sample()` step for the same reason
 - For D0CRT remove unused arg `random_state`
 - Add reproducibility and randomness tests

### CODE REVIEW

File: test/test_conditional_feature_importance.py [hunk 1]
```diff
@@ -74,10 +74,10 @@ def run_cfi(X, y, n_permutation, seed):
...
 ## tests cfi on different type of data
 parameter_exact = [
-    ("HiDim", 150, 200, 10, 0.0, 42, 1.0, np.inf, 0.0),
-    ("HiDim with noise", 150, 200, 10, 0.0, 42, 1.0, 10.0, 0.0),
-    ("HiDim with correlated noise", 150, 200, 10, 0.0, 42, 1.0, 10.0, 0.2),
-    ("HiDim with correlated features", 150, 200, 10, 0.2, 42, 1.0, np.inf, 0.0),
+    ("HiDim", 150, 10, 5, 0.0, 42, 1.0, np.inf, 0.0),
+    ("HiDim with noise", 150, 10, 5, 0.0, 42, 1.0, 10.0, 0.0),
+    ("HiDim with correlated noise", 150, 10, 5, 0.0, 42, 1.0, 10.0, 0.2),
+    ("HiDim with correlated features", 150, 10, 5, 0.2, 42, 1.0, np.inf, 0.0),
```
[the remark this answers is not in the document]
Comment (jpaillard) [reply test/test_conditional_feature_importance.py]: I returned to the previous test but allowed for a small tolerance: Asking for a perfect support recovery with n=150 p=200, with correlation and noise is ambitious. 
I test for the 10 important variables to be in the top 15 in importance returned by CFI.

[the remark this answers is not in the document]
Comment (lionelkusch) [reply test/test_conditional_feature_importance.py]: I made 3 categories of test:
- exact: which returns all the important variables
- partial: which returns all a part of it
 - failed: which doesn't succeed in finding them
 I would like to keep this spirit for all the tests.

File: test/test_conditional_feature_importance.py [hunk 2]
```diff
@@ -568,3 +569,135 @@ def test_groups_warning(self, data_generator):
             " number of features for which importance is computed: 4",
         ):
             cfi.importance(X, y)
+
+
+class TestCFIReproducibility:
```
Comment (bthirion) [inline test/test_conditional_feature_importance.py:574]: I would like to avoid classes in tests by all means, to make tests as flat as possible. I think that a class here is not bringing enough.

File: test/test_distilled_conditional_randomization_test.py [hunk 3]
```diff
@@ -354,3 +334,101 @@ def test_function_d0crt():
     assert len(sv) <= 10
     assert len(importances) == 10
     assert len(pvalues) == 10
+
+
+class TestD0CRTReproducibility:
```
Comment (bthirion) [inline test/test_distilled_conditional_randomization_test.py:339]: Same comment here

File: test/test_permutation_feature_importance.py [hunk 4]
```diff
@@ -99,3 +99,130 @@ def test_permutation_importance():
 
     importance_clf = vim_clf["importance"]
     assert importance_clf.shape == (X.shape[1],)
+
+
+class TestPFIReproducibility:
```
Comment (bthirion) [inline test/test_permutation_feature_importance.py:104]: and here too.

**Review by bthirion — COMMENTED** [review]
The PR looks good, but I would like to avoid classes in tests.
Thx !

File: test/test_conditional_feature_importance.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (jpaillard) [reply test/test_conditional_feature_importance.py]: CFI is probabilistic by nature; there are no guarantees that it will identify exactly the support 100% of the time. 
In addition, several factors make it even harder in the "exact" test: 
 1. n > p with p=150! 
 2. No CV is used for aggregating the importances over folds
 3. Only 10 permutations are used
 4. There is noise and correlation in the data 

I think the "exact" test is unrealistic.

**Comment by jpaillard** [comment]
I fixed it while keeping the option to pass `random_state=None` to the `ConditionalSampler`, which is necessary to generate randomness.

File: src/hidimstat/conditional_feature_importance.py [hunk 5]
```diff
@@ -1,8 +1,7 @@
 import numpy as np
 from joblib import Parallel, delayed
-from sklearn.base import check_is_fitted, clone, BaseEstimator
+from sklearn.base import BaseEstimator, check_is_fitted, clone
 from sklearn.metrics import root_mean_squared_error
-from sklearn.utils.validation import check_random_state
```
Comment (lionelkusch) [inline src/hidimstat/conditional_feature_importance.py:5]: Don't remove this line.
It's important to keep the check on the random generator from scikit learn.

Comment (lionelkusch) [inline src/hidimstat/conditional_feature_importance.py:5]: Don't remove this line.
It's important to keep the check on the random generator from scikit learn.

File: src/hidimstat/conditional_feature_importance.py [hunk 6]
```diff
@@ -104,8 +103,9 @@ def fit(self, X, y=None, groups=None, var_type="auto"):
         self : object
             Returns the instance itself.
         """
-        self.random_state = check_random_state(self.random_state)
         super().fit(X, None, groups=groups)
+        rng = np.random.default_rng(self.random_state)
```
Comment (lionelkusch) [inline src/hidimstat/conditional_feature_importance.py:107]:
```suggestion
        rng =  check_random_state(self.random_state)
```

File: src/hidimstat/conditional_feature_importance.py (hunk 5 above)
Comment (jpaillard) [reply src/hidimstat/conditional_feature_importance.py:5]: done

File: test/test_conditional_feature_importance.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (jpaillard) [reply test/test_conditional_feature_importance.py]: It's a question of what we call a modification of the behavior. IMO, modifying the seeding (done in this PR) of the method (done in this PR) does not change its behavior, and tests should not be sensitive to this. Otherwise, a simple change in numpy's generator, completely independent of the method, could make the test fail.

[the remark this answers is not in the document]
Comment (bthirion) [reply test/test_conditional_feature_importance.py]: A test that fails with half of the seeds is clearly a bad test. But I'd be more strict: any test that fails with 1% of the seeds is already bad.

[the remark this answers is not in the document]
Comment (bthirion) [reply test/test_conditional_feature_importance.py]: It depends on what exactly you want to test, but the straightforward solution is to make the simulation simple enough that it always work. Do you have any particular test in mind ?

[the remark this answers is not in the document]
Comment (lionelkusch) [reply test/test_conditional_feature_importance.py]: I just want to test the main functionality of the library, which is the detection of variable importance.
Some methods for this case don't have theoretical boundaries. This means that it is not possible to have a percentage of the seed which is not working or not.
However, if it's working for some data, it still should work. This is why having a test works for 50% of the seeds, but testing only one is enough to check if the method does not change behaviour because some small modifications of the code won't break the test due to a tolerance of a large number of seeds. However, if it's working for only one specific seed, in this case, any modification of the code has a high chance of breaking the test for not a good reason.
If we consider that the failing of 1% of seeds means that we can't test some methods because there is a theoretical boundary. For continuous integration, we need to have tests to identify if the main functionality of the function is constant over time.

File: test/test_conditional_feature_importance.py [hunk 7]
```diff
@@ -89,12 +90,12 @@ def run_cfi(X, y, n_permutation, seed):
 @pytest.mark.parametrize("n_permutation, cfi_seed", [(10, 0)], ids=["default_cfi"])
 def test_linear_data_exact(data_generator, n_permutation, cfi_seed):
     """Tests the method on linear cases with noise and correlation"""
-    X, y, important_features, _ = data_generator
+    X, y, important_features, not_important_features = data_generator
```
Comment (bthirion) [inline test/test_conditional_feature_importance.py:93]: It looks like this change is not needed

File: test/test_distilled_conditional_randomization_test.py [hunk 8]
```diff
@@ -50,7 +46,6 @@ def test_dcrt_lasso_screening(generate_regation_dataset):
     d0crt_no_screening = D0CRT(
         estimator=LassoCV(n_jobs=1),
         screening=False,
-        random_state=2024,
```
Comment (bthirion) [inline test/test_distilled_conditional_randomization_test.py:53]: Why are you removing random_states in these functions ?

File: test/test_conditional_feature_importance.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (lionelkusch) [reply test/test_conditional_feature_importance.py]: I want to have tests for high dimension cases because it's important to check for this type of data.

I take the example of knockoff because it's the most problematic on this topic.
If you consider that this is not true, can you provide a mathematical description of the rate of false true positives?

[the remark this answers is not in the document]
Comment (bthirion) [reply test/test_conditional_feature_importance.py]: No, tests should not include cases that are too difficult (in the sense of being hardly identifiable). That's not their purpose. And we don't want to have flaky tests.

[the remark this answers is not in the document]
Comment (lionelkusch) [reply test/test_conditional_feature_importance.py]: I open the issue #375 on this topic. 
@jpaillard, can you reverse the test for the moment?
If the test for "HiDim with correlated noise" doesn't pass, add it to `parameter_partial` and adjust the missing rank if it's necessary.

**Review by bthirion — APPROVED** [review]
LGTM, thx.

**Comment by jpaillard** [comment]
For sklearn estimators, we can simply use `np.random.RandomState(rng.bit_generator)`, that way we will have rng that are generated using the `SeedSequences` you suggested without needing to add a function.

File: test/test_distilled_conditional_randomization_test.py [hunk 9]
```diff
@@ -344,13 +330,135 @@ def test_dcrt_invalid_lasso_screening(generate_regression_dataset):
...
+        n_features=10,
+        noise=0.2,
+    )
     sv, importances, pvalues = d0crt(LassoCV(n_jobs=1), X, y)
     assert len(sv) <= 10
     assert len(importances) == 10
     assert len(pvalues) == 10
 
 
+@pytest.fixture(scope="module")
```
[the remark this answers is not in the document]
Comment (jpaillard) [reply test/test_distilled_conditional_randomization_test.py:344]: done

File: test/test_permutation_feature_importance.py [hunk 10]
```diff
@@ -99,3 +99,112 @@ def test_permutation_importance():
 
     importance_clf = vim_clf["importance"]
     assert importance_clf.shape == (X.shape[1],)
+
+
+@pytest.fixture(scope="module")
```
[the remark this answers is not in the document]
Comment (jpaillard) [reply test/test_permutation_feature_importance.py:104]: done

File: src/hidimstat/desparsified_lasso.py [hunk 11]
```diff
@@ -447,7 +458,9 @@ def _compute_all_residuals(
     return Z, precision_diagonal
 
 
-def _compute_residuals(X, id_column, alpha, gram, max_iteration=5000, tolerance=1e-3):
+def _compute_residuals(
```
[the remark this answers is not in the document]
Comment (jpaillard) [reply src/hidimstat/desparsified_lasso.py:461]: done

File: src/hidimstat/distilled_conditional_randomization_test.py [hunk 12]
```diff
@@ -120,8 +120,8 @@ def __init__(
         joblib_verbose=0,
         fit_y=False,
         scaled_statistics=False,
-        random_state=None,
         reuse_screening_model=True,
+        random_state=None,
```
[the remark this answers is not in the document]
Comment (jpaillard) [reply src/hidimstat/distilled_conditional_randomization_test.py:124]: done

File: src/hidimstat/gaussian_knockoff.py [hunk 13]
```diff
@@ -2,6 +2,8 @@
 
 import numpy as np
 
+from hidimstat._utils.utils import check_random_state
+
 
 def gaussian_knockoff_generation(X, mu, sigma, seed=None, tol=1e-14):
```
[the remark this answers is not in the document]
Comment (jpaillard) [reply src/hidimstat/gaussian_knockoff.py:8]: done

File: test/test_conditional_feature_importance.py [hunk 14]
```diff
@@ -75,10 +76,10 @@ def run_cfi(X, y, n_permutation, seed):
 ##############################################################################
 ## tests cfi on different type of data
 parameter_exact = [
-    ("HiDim", 150, 200, 10, 0.0, 42, 1.0, np.inf, 0.0),
-    ("HiDim with noise", 150, 200, 10, 0.0, 42, 1.0, 10.0, 0.0),
-    ("HiDim with correlated noise", 150, 200, 10, 0.0, 42, 1.0, 10.0, 0.2),
-    ("HiDim with correlated features", 150, 200, 10, 0.2, 42, 1.0, np.inf, 0.0),
+    ("HiDim", 150, 20, 10, 0.0, 42, 1.0, np.inf, 0.0),
```
[the remark this answers is not in the document]
Comment (jpaillard) [reply test/test_conditional_feature_importance.py:79]: I reverted the changes, but for the record, this PR is the example of why we don't want such unstable tests. It leads to the failure of the CI because of the expected stochastic fluctuations of the method, and not due to changes in the core logic of the method.

File: test/test_conditional_feature_importance.py [hunk 15]
```diff
@@ -95,7 +96,7 @@ def test_linear_data_exact(data_generator, n_permutation, cfi_seed):
     # check that importance scores are defined for each feature
     assert importance.shape == (X.shape[1],)
     # check that important features have the highest importance scores
-    assert np.all([int(i) in important_features for i in np.argsort(importance)[-10:]])
+    assert np.all([int(i) in np.argsort(importance)[-10:] for i in important_features])
```
[the remark this answers is not in the document]
Comment (jpaillard) [reply test/test_conditional_feature_importance.py:99]: same as above

File: test/test_conditional_feature_importance.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (bthirion) [reply test/test_conditional_feature_importance.py]: Indeed !

File: test/test_distilled_conditional_randomization_test.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (lionelkusch) [reply test/test_distilled_conditional_randomization_test.py:345]: Can you add the default argument of D0CRT to the fixture; this will help to read the tests?

File: test/test_permutation_feature_importance.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (lionelkusch) [reply test/test_permutation_feature_importance.py:104]: Can you add the default argument of PFI to the fixture; this will help to read the tests?

Can you use the same order of the test than for the test?
- repeatibility
- with_None
- with_integer
- with_RNG

File: src/hidimstat/_utils/utils.py [hunk 16]
```diff
@@ -25,3 +30,46 @@ def _check_vim_predict_method(method):
...
+
+def check_random_state(seed):
+    """
+    Modified version of sklearn's check_random_state using np.random.Generator.
+    This is based on the implementation of check_random_state of sciktilearn:
+    https://github.com/scikit-learn/scikit-learn/blob/25dee604bae18205b01548348388baf7a1cdfe0e/sklearn/utils/validation.py#L1488
+
+    Parameters
+    ----------
+    seed : None, int, Generator or instance of RandomState
```
Comment (lionelkusch) [inline src/hidimstat/_utils/utils.py:43]:
```suggestion
    seed : None, int or Generator
```

**Review by lionelkusch — COMMENTED** [review]
Thank you for all the small modifications.

I just ask you one modification of the test before that this PR is ready to be merged.

I have a question: why do you remove all the random_state parameters in the tests?

**Comment by jpaillard** [comment]
The test `test_ensemble_clustered_inference_temporal_data` was also unstable and blocking for the reason discussed above, even though we only touched the random_state. 
I had to reduce n=200, p=2000 --> n=200, p=400

**Comment by jpaillard** [comment]
Thx for spotting, 
 - Conditional vs Marginal Importance on the XOR dataset: it's not far from the chance level, the message remains the same. When using a single random_state=0 in all the examples, the boxplot overlaps more clearly with the chance level. 
 -  Pitfalls of Permutation Feature Importance (PFI) on the California Housing Dataset: the issue was introduced in #344 (now also in main), in which inconsistent KFold objects were used throughout the example. I fixed it, and it matches the original example again. 
 - For the iris dataset, I'm not so surprised; it seems that petal features are not so important.

**Comment by lionelkusch** [comment]
I find an error in the following example:
Controlled multiple variable selection on the Wisconsin breast cancer dataset
image
Text: It is also interesting to note that some of the selected variables (with the high KO statistic (e.g., worst radius, worst area, mean concave points) are also variables with the largest Lasso coefficients.

File: test/test_conditional_feature_importance.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (jpaillard) [reply test/test_conditional_feature_importance.py]: done

File: src/hidimstat/distilled_conditional_randomization_test.py
Comment (lionelkusch) [inline src/hidimstat/distilled_conditional_randomization_test.py:103]: Add a sentences about the management of the random_state of the model.

File: src/hidimstat/gaussian_knockoff.py [hunk 17]
```diff
@@ -20,17 +22,13 @@ def gaussian_knockoff_generation(X, mu, sigma, seed=None, tol=1e-14):
...
         The original design matrix.
-
     mu : 1D ndarray (n_features, )
         A vector of empirical mean values.
-
     sigma : 2D ndarray (n_samples, n_features)
         The empirical covariance matrix.
-
-    seed : int, optional
+    random_state : int or None, optional
```
Comment (lionelkusch) [inline src/hidimstat/gaussian_knockoff.py:29]:
```suggestion
    random_state : int or None, default=None
```

**Review by lionelkusch — CHANGES_REQUESTED** [review]
Few modification on the docstring
